earth_enterprise/src/fusion/autoingest/AssetGenDaemonCpp.pl

```diff
@@ -414,18 +412,21 @@ ()
     void AddConfig(DOMElement *parent, const $config &config);
 }
 
-khRefGuard<${name}AssetImplD>
+std::shared_ptr<${name}AssetImplD>
 ${name}AssetImplD::Load(const std::string &boundref)
 {
-    khRefGuard<${name}AssetImplD> result;
-
     // make sure the base class loader actually instantiated one of me
     // this should always happen, but there are no compile time guarantees
-    result.dyncastassign(${name}AssetImpl::Load(boundref));
-    if (!result) {
-        AssetThrowPolicy::FatalOrThrow(
-            "Internal error: ${name}AssetImplD loaded wrong type for " +
-            boundref);
+    auto loaded = ${name}AssetVersionImpl::Load(boundref);
+    std::shared_ptr<${name}AssetImplD> result =
+        std::dynamic_pointer_cast<${name}AssetImplD>(loaded);
+
+    if (result == nullptr) {
+        std::string error {"Internal error: "};
+        if (loaded == nullptr)
```

Comment:
Very minor: might be better to have two completely separate error messages here (and in the similar code in other places) - one error for the base not loading and a separate error for the conversion not working.